node/network/availability-distribution/src/lib.rs

```diff
@@ -0,0 +1,1852 @@
+
+	/// The set of validators.
+	validators: Vec<ValidatorId>,
+
+	/// If this node is a validator, note the index in the validator set.
+	validator_index: Option<ValidatorIndex>,}
+
+#[derive(Debug, Clone, Default)]
+struct PerRelayParent {
+	/// Set of `K` ancestors for this relay parent.
+	ancestors: Vec<Hash>,
+}
+
+impl ProtocolState {
+	// collects the relay_parents plus their ancestors
+	fn extend_with_ancestors<'a>(
+		&'a self,
+		relay_parents: impl IntoIterator<Item = &'a Hash> + 'a,
+	) -> HashSet<Hash> {
+		relay_parents
+			.into_iter()
+			.map(|relay_parent| {
+				self.per_relay_parent
+					.get(relay_parent)
+					.into_iter()
+					.map(|per_relay_parent| per_relay_parent.ancestors.iter().cloned())
+					.flatten()
+					.chain(iter::once(relay_parent.clone()))
+			})
+			.flatten()
+			.collect::<HashSet<Hash>>()
+	}
+
+	/// Unionize all cached entries for the given relay parents and it's ancestors ancestors.
+	/// Ignores all non existent relay parents, so this can be used directly with a peers view.
+	/// Returns a map from candidate hash -> receipt
+	fn cached_live_candidates_unioned<'a>(
+		&'a self,
+		relay_parents: impl IntoIterator<Item = &'a Hash> + 'a,
+	) -> HashMap<Hash, CommittedCandidateReceipt> {
+		let relay_parents_and_ancestors = self.extend_with_ancestors(relay_parents);
+		relay_parents_and_ancestors
+			.into_iter()
+			.filter_map(|relay_parent_or_ancestor| self.receipts.get(&relay_parent_or_ancestor))
+			.map(|receipt_set| receipt_set.into_iter())
+			.flatten()
+			.map(|(receipt_hash, receipt)| (receipt_hash.clone(), receipt.clone()))
+			.collect::<HashMap<Hash, CommittedCandidateReceipt>>()
+	}
+
+	async fn add_relay_parent<Context>(
+		&mut self,
+		ctx: &mut Context,
+		relay_parent: Hash,
+		validators: Vec<ValidatorId>,
+		validator_index: Option<ValidatorIndex>,
+	) -> Result<()>
+	where
+		Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
+	{
+		let candidates =
+			query_live_candidates(ctx, self, std::iter::once(relay_parent.clone())).await?;
+
+		// register the relation of relay_parent to candidate..
+		// ..and the reverse association.
+		for (relay_parent_or_ancestor, (receipt_hash, receipt)) in candidates.clone() {
+			self
+				.reverse
+				.insert(receipt_hash.clone(), relay_parent_or_ancestor.clone());
+			let per_candidate = self.per_candidate.entry(receipt_hash.clone())
+				.or_default();
+			per_candidate.validator_index = validator_index.clone();
+			per_candidate.validators = validators.clone();
+
+			self
+				.receipts
+				.entry(relay_parent_or_ancestor)
+				.or_default()
+				.insert((receipt_hash, receipt));
+		}
+
+		// collect the ancestors again from the hash map
+		let ancestors = candidates
+			.iter()
+			.filter_map(|(ancestor_or_relay_parent, _receipt)| {
+				if ancestor_or_relay_parent == &relay_parent {
+					None
+				} else {
+					Some(ancestor_or_relay_parent.clone())
+				}
```

Comment:
feels like a `filter(ancestor_or_relay_parent != &relay_parent)` would be simpler here

Reply:
The main reason is, that I can go from `&T -> T` by means of `Clone` at the same time as filtering.